### PR TITLE
Fix `finite_opt_saving.py` and `modified_opt_saving.py`

### PR DESCRIPTION
**Pull request by HumphreyYang**
Hi @orectique, @jstac, and @Smit-create,

This PR fixes small bugs in `finite_opt_saving.py`.

Although this code can be run and draw the same plots without error, the HPI is significantly slower compared to the graph in the book. One reason is that `itertools.product` is incompatible with Numba, and I have to homebrew an easy Numba-compatible version of `cartesian_product`. I have tried other Numba-compatible implementations of the Cartesian product, but the speed is roughly the same. 

Figure_1

There might be other factors causing HPI to be slower.

### CODE REVIEW

**Comment by jstac** [comment]
Thanks for picking this up @HumphreyYang .

Your version is clever, but I think you can eliminate `cartesian_product` and just use a nested loop in the function.

The only reason I used `product` in the Julia version of the code was to fit the code snippet onto the page in the LaTeX book :-)

So you don't need to exactly follow that structure.  Would you mind to try the nested for loop instead?

**Comment by Smit-create** [comment]
Thanks, @HumphreyYang. These improvements look great to me. I didn't try these things at the time of converting as I focused on the thought that people can easily migrate between Julia/Python code and so kept line-by-line changes from Julia to Python and didn't focus on optimization. Surely, there are many scopes of optimizing.

**Comment by HumphreyYang** [comment]
Many thanks @jstac and @Smit-create,

Sorry about the delay, as I was playing with different setups. I was able to achieve this plot that we aim for using the optimized code:

Screenshot 2023-08-07 215756

## Note: 

The speed gain is only achieved when the code is run for the second time and onwards. This relates to the `cache` option in `njit` as it reduces the overhead in the run.

**Comment by spulick** [comment]
Thanks for the fix @HumphreyYang. 

Shall I leave this PR open till we fix all the scripts?

**Comment by HumphreyYang** [comment]
Many thanks @orectique,

Just bear with me for another 10 minutes. The `modified_optimal_saving` should be an easy fix : ) Let's put them together in this PR.

**Comment by HumphreyYang** [comment]
Hi @jstac,

Both `modified_opt_saving.py` and `finite_opt_saving.py` are updated in this PR. @orectique please kindly merge once we have the approval from @jstac : )

Many thanks in advance.

**Comment by jstac** [comment]
Great work @HumphreyYang , many thanks.

**Review by jstac — APPROVED** [review]

**Comment by spulick** [comment]
Thank you, @jstac @HumphreyYang @Smit-create

**Pull request merged**
